api/oc_push.c

```diff
@@ -0,0 +1,2605 @@
+		oc_list_remove(pushd_rsc_rep_list, pushd_rsc_rep);
+		oc_memb_free(&pushd_rsc_rep_instance_memb, pushd_rsc_rep);
+	}
+	else
+	{
+		OC_PUSH_ERR("can't find resource representation for pushed resource (%s)...", oc_string(*uri));
+		return;
+	}
+
+	if (pushd_rsc)
+	{
+		/* step 3. remove pushed Resource from `app_resources` */
+		OC_PUSH_DBG("purge pushed resource (%s)...", oc_string(*uri));
+		oc_delete_resource(pushd_rsc);
+	}
+	else
+	{
+		OC_PUSH_ERR("can't find pushed resource (%s)...", oc_string(*uri));
+		return;
+	}
+
+	return;
+}
+
+
+
+/*
+ * @brief				create app resource and `oc_pushd_rsc_t` corresponding to pushed resource
+ *
+ * @param recv_obj	receiver object that points pushed resource
+ * @param resource	Push Receiver resource
+ */
+static bool _create_pushd_rsc(oc_recv_t *recv_obj, oc_resource_t *resource)
+{
+	bool result = true;
+
+	/* create Push Receiver Resource */
+	oc_resource_t *pushd_rsc = oc_new_resource("Pushed Resource", oc_string(recv_obj->receiveruri), 1, resource->device);
+
+	if (pushd_rsc)
+	{
+		/*
+		 * XXX, if a resource binds empty resource type (""), when a client retrieve this it may receive weird value...
+		 */
+		oc_resource_bind_resource_type(pushd_rsc, " ");
+		oc_resource_bind_resource_interface(pushd_rsc, OC_IF_RW | OC_IF_BASELINE);
+		oc_resource_set_default_interface(pushd_rsc, OC_IF_RW);
+		/*
+		 * initially this resource should not be discoverable...
+		 * once any resource is pushed to this resource, it will be discoverable...
+		 */
+		oc_resource_set_discoverable(pushd_rsc, true);
+
+		oc_resource_set_request_handler(pushd_rsc, OC_GET, get_pushd_rsc, NULL);
+		oc_resource_set_request_handler(pushd_rsc, OC_POST, post_pushd_rsc, NULL);
+		/*
+		 * when this pushed resource is deleted.. delete corresponding "receiver" object from receivers array of push receiver resource
+		 * => this is done in delete_pushrecv() (delete handler of pushreceiver resource)
+		 */
+
+		if (!oc_add_resource(pushd_rsc))
+			result = false;
+
+		/* create resource representation container for this resource */
+		oc_pushd_rsc_rep_t *pushd_rsc_rep_instance = (oc_pushd_rsc_rep_t *)oc_memb_alloc(&pushd_rsc_rep_instance_memb);
+		if (pushd_rsc_rep_instance) {
+			pushd_rsc_rep_instance->resource = pushd_rsc;
+			pushd_rsc_rep_instance->rep = NULL;
+			oc_list_add(pushd_rsc_rep_list, pushd_rsc_rep_instance);
+		}
+		else
+		{
+			OC_PUSH_ERR("oc_memb_alloc() error!");
+			result = false;
+		}
+	}
+	else
+	{
+		OC_PUSH_ERR("oc_new_resource() error!");
+		result = false;
+	}
+
+	return result;
+}
+
+
+
+
+/*
+ * @brief						remove receiver object array in `recv_obj_list`,
+ * 								and app resource pointed by `receiveruri` of each receivre object in the array
+ *
+ * @param recv_obj_list		receiver object array
+ * @param device_index		index of device where the Push Resource resides
+ */
+static void _purge_recv_obj_list(oc_recvs_t *recvs_instance)
+{
+	/*
+	 * TODO4ME rewrite this func (refer to oc_free_rep())
```

Comment:
Is this (and other) TODO4ME still relevant? :-)

Reply:
it's strange, in the latest commit that comment was already removed.
Are you seeing old commit maybe ??
(I have not rebased yet to identify code modification reflecting review comment easily)

Reply:
Yeah, this one is gone, though they are still 2 TODO4ME in the code, if I'm not being deceived again.

Reply:
oh yes, I know that.
I leave them intentionally as future enhancement works. 
Current implementation has already passed CTT TCs and those TODOS does not affect spec satisfaction